flowmachine/flowmachine/features/subscriber/home_location_monthly.py

```diff
@@ -0,0 +1,198 @@
+from typing import List, Union
+from flowmachine.core.spatial_unit import AnySpatialUnit
+from datetime import datetime, timedelta
+from flowmachine.core.query import Query
+from dateutil.relativedelta import relativedelta
+from dateutil.rrule import rrule, DAILY
+from copy import copy
+
+from flowmachine.features.subscriber.daily_location import daily_location
+from flowmachine.features.subscriber.modal_location import ModalLocation
+from flowmachine.features.subscriber.unique_active_subscribers import (
+    UniqueActiveSubscribers,
+    ActiveSubscribers,
+)
+from flowmachine.features.subscriber.last_location import LastLocation
+from flowmachine.features.subscriber.location_visits import LocationVisits
+
+
+class HomeLocationMonthly(Query):
+
+    """For each subscriber, if their modal last location was their last location on more than X days this month
+    (choose X=15 here), then choose this as their home location. If not, but it was their last location on more than Y
+     days this month (Y<X; choose Y=10 here) AND it is the same as their “reference location” (in our case reference
+     location is modal location and last location on more than X days last month), then choose this as their home
+     location. Otherwise set their home location to “unlocatable” (a special value, not NULL)."""
+
+    def __init__(
+        self,
+        window_start: Union[str, datetime],
```

Comment:
Rather than taking a date range and constructing the sub-queries internally, I think it would be better to take three queries (ModalLocation, LocationVisits and ActiveSubscribers) as parameters, because:
- the date ranges aren't necessarily the same for each of these queries (e.g. in the original query plan, modal location is over 40 days, but the daily location visit counts are over one month)
- this would make the query more flexible (e.g. if we want to use an additional rolling filter before calculating the modal location, that could be done without changing the implementation of this query)
- the ActiveSubscribers query here is just a subscriber subset - passing the query in directly as a 'subscriber_subset' parameter would mean users could choose to use an entirely different subscriber subset if preferred.

Reply:
Actually, in this situation the ActiveSubscribers query wouldn't be needed in this query at all - the subscriber subset would be applied to each of the other queries before passing them into this query.